src/Data/Singletons/Promote.hs

```diff
@@ -391,15 +436,25 @@ promoteMethod m_subst sigs_map (meth_name, meth_rhs) = do
     lookup_meth_ty :: PrM ([DKind], DKind)
     lookup_meth_ty = case Map.lookup meth_name sigs_map of
       Nothing -> do
-        mb_info <- dsReify proName
-        case mb_info of
-          Just (DTyConI (DOpenTypeFamilyD (DTypeFamilyHead _ tvbs mb_res_ki _)) _)
-            -> let arg_kis = map (default_to_star . extractTvbKind) tvbs
-                   res_ki  = default_to_star (resultSigToMaybeKind mb_res_ki)
-               in return (arg_kis, res_ki)
-          _ -> fail $ "Cannot find type annotation for " ++ show proName
+        let mk_meth_ty = extract_meth_ty (dsReify proName)
+                     <|> extract_meth_ty (dsReifyTypeNameInfo proName)
+                     -- See Note [Using dsReifyTypeNameInfo when promoting instances]
```

Comment:
Indeed, it's much simpler now. I still need to bump the `th-desugar` submodule, though, since I'm using [`mkTypeNameWithLocals`](https://github.com/goldfirere/th-desugar/pull/62#issuecomment-314827010).